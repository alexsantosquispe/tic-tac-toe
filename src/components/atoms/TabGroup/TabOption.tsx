import cn from 'clsx';
import type { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';

interface TabOptionProps<T> {
  id: T;
  label?: string;
  icon?: ReactNode;
  isSelected?: boolean;
  onSelect: (optionId: T) => void;
  isDefault?: boolean;
}

const TabOption = <T,>({
  id,
  label,
  icon,
  isSelected = false,
  onSelect,
  isDefault = true
}: TabOptionProps<T>) => {
  return (
    <button
      aria-label={`${id} button`}
      className={twMerge(
        'relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white',
        cn({
          'text-white dark:text-black': isSelected && isDefault,
          'text-white': isSelected && !isDefault
        })
      )}
      onClick={() => onSelect(id)}
    >
      {icon && icon}
      {label && <pre>{label}</pre>}
    </button>
  );
};

export default TabOption;
