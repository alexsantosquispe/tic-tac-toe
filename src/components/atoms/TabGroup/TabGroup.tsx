import cn from 'clsx';
import type { ReactNode } from 'react';
import { twMerge } from 'tailwind-merge';
import TabOption from './TabOption';

type OptionType<T> = {
  id: T;
  label?: string;
  icon?: ReactNode;
};

interface TabGroupProps<T> {
  options: OptionType<T>[];
  optionIdSelected: T;
  onSelectOption: (optionId: T) => void;
  isDefault?: boolean;
}

export const TabGroup = <T extends string | number | symbol>({
  options,
  optionIdSelected,
  onSelectOption,
  isDefault = true
}: TabGroupProps<T>) => {
  const transformMap: Record<T, string> = options.reduce(
    (acc, option, index) => {
      acc[option.id] = `translate-x-[${index * 2}rem]`;
      return acc;
    },
    {} as Record<T, string>
  );
  const currentTransform = transformMap[optionIdSelected];

  return (
    <div className="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20">
      <div
        data-testid="selected-indicator"
        className={twMerge(
          'absolute top-1 left-1 h-7 w-7 rounded-md shadow',
          'transform transition-transform duration-200 ease-in-out',
          currentTransform,
          cn({
            'bg-rose-600': !isDefault,
            'bg-black dark:bg-white': isDefault
          })
        )}
      />

      {options.map(({ id, label, icon }) => {
        const isSelected = id === optionIdSelected;
        return (
          <TabOption
            key={id as string}
            id={id}
            label={label}
            icon={icon}
            isSelected={isSelected}
            onSelect={onSelectOption}
            isDefault={isDefault}
          />
        );
      })}
    </div>
  );
};
