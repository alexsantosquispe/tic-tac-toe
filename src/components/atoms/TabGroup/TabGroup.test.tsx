import { fireEvent, render, screen } from '@testing-library/react';

import { TabGroup } from './TabGroup';

describe('TabGroup', () => {
  const onSelectOptionMock = jest.fn();
  const props = {
    options: [
      { id: '1', label: 'Option 1' },
      { id: '2', label: 'Option 2' },
      { id: '3', label: 'Option 3' }
    ],
    optionIdSelected: '1',
    onSelectOption: onSelectOptionMock
  };

  it('should render the component correctly', () => {
    let component = render(<TabGroup {...props} />);

    expect(component).toMatchSnapshot();

    component = render(<TabGroup {...props} isDefault={false} />);

    expect(component).toMatchSnapshot();
  });

  it('should call onSelectOption when a tab is clicked', () => {
    render(<TabGroup {...props} />);

    const option2 = screen.getByText('Option 2');

    fireEvent.click(option2);

    expect(onSelectOptionMock).toHaveBeenCalledTimes(1);

    expect(onSelectOptionMock).toHaveBeenCalledWith('2');
  });
});
