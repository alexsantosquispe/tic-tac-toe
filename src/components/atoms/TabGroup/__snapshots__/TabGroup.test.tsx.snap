// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TabGroup should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
      >
        <div
          class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-black dark:bg-white"
          data-testid="selected-indicator"
        />
        <button
          aria-label="1 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer text-white dark:text-black"
        >
          <pre>
            Option 1
          </pre>
        </button>
        <button
          aria-label="2 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 2
          </pre>
        </button>
        <button
          aria-label="3 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 3
          </pre>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
    >
      <div
        class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-black dark:bg-white"
        data-testid="selected-indicator"
      />
      <button
        aria-label="1 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer text-white dark:text-black"
      >
        <pre>
          Option 1
        </pre>
      </button>
      <button
        aria-label="2 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <pre>
          Option 2
        </pre>
      </button>
      <button
        aria-label="3 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <pre>
          Option 3
        </pre>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TabGroup should render the component correctly 2`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
      >
        <div
          class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-black dark:bg-white"
          data-testid="selected-indicator"
        />
        <button
          aria-label="1 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer text-white dark:text-black"
        >
          <pre>
            Option 1
          </pre>
        </button>
        <button
          aria-label="2 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 2
          </pre>
        </button>
        <button
          aria-label="3 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 3
          </pre>
        </button>
      </div>
    </div>
    <div>
      <div
        class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
      >
        <div
          class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-rose-600"
          data-testid="selected-indicator"
        />
        <button
          aria-label="1 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer dark:text-white text-white"
        >
          <pre>
            Option 1
          </pre>
        </button>
        <button
          aria-label="2 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 2
          </pre>
        </button>
        <button
          aria-label="3 button"
          class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
        >
          <pre>
            Option 3
          </pre>
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="relative flex items-center gap-1 rounded-lg border border-neutral-200 p-1 dark:border-white/20"
    >
      <div
        class="absolute top-1 left-1 h-7 w-7 rounded-md shadow transform transition-transform duration-200 ease-in-out translate-x-[0rem] bg-rose-600"
        data-testid="selected-indicator"
      />
      <button
        aria-label="1 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs hover:cursor-pointer dark:text-white text-white"
      >
        <pre>
          Option 1
        </pre>
      </button>
      <button
        aria-label="2 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <pre>
          Option 2
        </pre>
      </button>
      <button
        aria-label="3 button"
        class="relative z-10 flex h-7 w-7 items-center justify-center text-xs text-neutral-600 hover:cursor-pointer dark:text-white"
      >
        <pre>
          Option 3
        </pre>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
